### PR TITLE
shipit_code_coverage: Wait for the build to be ingested by the coverage service and request the backend to generate and cache the data for all changesets in the push

### LINKED ISSUE #615: shipit_code_coverage: Populate coverage by changeset cache when uploading reports to coverage service
See #614.

### PR DESCRIPTION
**Pull request by marco-c**
Fixes #615.

### CODE REVIEW

File: src/shipit_code_coverage/shipit_code_coverage/uploader.py [hunk 1]
```diff
@@ -57,3 +58,11 @@ def codecov(data, commit_sha, token, flags=None):
...
     if r.status_code != requests.codes.ok:
         raise Exception('Failure to upload data to S3. Response [%s]: %s' % (r.status_code, r.text))
+
+
+def codecov_wait(commit):
+    def check_codecov_job():
+        r = requests.get('https://codecov.io/api/gh/marco-c/gecko-dev/commit/%s' % commit)
+        return True if r.json()['commit']['totals'] is not None else False
+
+    return utils.wait_until(check_codecov_job, 30) is not None
```
Comment (armenzg) [inline src/shipit_code_coverage/shipit_code_coverage/uploader.py]: Is that 30 seconds? or minutes?

File: src/shipit_code_coverage/shipit_code_coverage/codecov.py [hunk 2]
```diff
@@ -238,3 +238,22 @@ def go(self):
...
+        # data, which will be cached.
+        r = requests.get('https://hg.mozilla.org/mozilla-central/json-rev/%s' % self.revision)
+        push_id = int(r.json()['pushid'])
+
+        r = requests.get('https://hg.mozilla.org/mozilla-central/json-pushes?startID=%s&endID=%s' % (push_id - 1, push_id))
+        data = r.json()
+        changesets = data[str(push_id)]['changesets']
+
+        for changeset in changesets:
+            requests.get('https://uplift.shipit.staging.mozilla-releng.net/coverage/changeset/%s' % changeset)
```
Comment (armenzg) [inline src/shipit_code_coverage/shipit_code_coverage/codecov.py]: What happens if the changeset API fetches from codecov and something fails?
Do we cache incorrect information?
Do we have a mechanism to clear a cache if we were to need it?

These questions might be steaming from an improper understanding. Thanks for your patience if so!

File: src/shipit_code_coverage/shipit_code_coverage/codecov.py [hunk 3]
```diff
@@ -238,3 +238,22 @@ def go(self):
         with ThreadPoolExecutorResult(max_workers=2) as executor:
             executor.submit(lambda: uploader.coveralls(output))
             executor.submit(lambda: uploader.codecov(output, commit_sha, self.codecov_token))
+
+        logger.info('Waiting for build to be injested by Codecov...')
```
[the remark this answers is not in the document]
Comment (marco-c) [reply src/shipit_code_coverage/shipit_code_coverage/codecov.py]: Done in 2e81e1a9.

File: src/shipit_code_coverage/shipit_code_coverage/codecov.py [hunk 4]
```diff
@@ -238,3 +238,22 @@ def go(self):
...
+        # Wait until the build has been injested by Codecov.
+        if uploader.codecov_wait(commit_sha):
```
[the remark this answers is not in the document]
Comment (marco-c) [reply src/shipit_code_coverage/shipit_code_coverage/codecov.py]: Done in bdc96a3b.

Yes, there could be a delay. `codecov_wait` only waits 30 minutes before returning (so if it takes longer, the data won't be cached).

File: src/shipit_code_coverage/shipit_code_coverage/codecov.py (hunk 2 above)
Comment (marco-c) [reply src/shipit_code_coverage/shipit_code_coverage/codecov.py]: If the report is not ingested by codecov, we will not cache the data at all.
I think there is a way to clear the cache, but I don't have permissions on Heroku to do it (I will at some point).

File: src/shipit_code_coverage/shipit_code_coverage/uploader.py (hunk 1 above)
Comment (marco-c) [reply src/shipit_code_coverage/shipit_code_coverage/uploader.py]: 30 minutes

File: src/shipit_code_coverage/shipit_code_coverage/codecov.py [hunk 5]
```diff
@@ -238,3 +238,22 @@ def go(self):
...
+            logger.info('codecov.io took too much time to injest data.')
+
+        # Get pushlog and ask the backend to generate the coverage by changeset
+        # data, which will be cached.
+        r = requests.get('https://hg.mozilla.org/mozilla-central/json-rev/%s' % self.revision)
+        push_id = int(r.json()['pushid'])
+
+        r = requests.get('https://hg.mozilla.org/mozilla-central/json-pushes?startID=%s&endID=%s' % (push_id - 1, push_id))
+        data = r.json()
+        changesets = data[str(push_id)]['changesets']
```
[the remark this answers is not in the document]
Comment (marco-c) [reply src/shipit_code_coverage/shipit_code_coverage/codecov.py]: Done in 2878c46c.

**Pull request merged**